<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="RNA-seek is an open-source, reproducible, and scalable best practices pipeline for analyzing RNA-sequencing data. "><meta name=author content="Skyler Kuhn"><link rel=icon href=../../assets/favicon/favicon.ico><meta name=generator content="mkdocs-1.1.2, mkdocs-material-7.2.3"><title>Getting started - RNA-seek Documentation</title><link rel=stylesheet href=../../assets/stylesheets/main.f7f47774.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.3f5d1f46.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback"><style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <script>function __prefix(e){return new URL("../..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script> <script>var palette=__get("__palette");if(null!==palette&&"object"==typeof palette.color)for(var key in palette.color)document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#1-introduction class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="RNA-seek Documentation" class="md-header__button md-logo" aria-label="RNA-seek Documentation" data-md-component=logo> <img src=../../assets/icons/doc-book.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> RNA-seek Documentation </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Getting started </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M7 10a2 2 0 0 1 2 2 2 2 0 0 1-2 2 2 2 0 0 1-2-2 2 2 0 0 1 2-2m10-3a5 5 0 0 1 5 5 5 5 0 0 1-5 5H7a5 5 0 0 1-5-5 5 5 0 0 1 5-5h10M7 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3h10a3 3 0 0 0 3-3 3 3 0 0 0-3-3H7z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=red data-md-color-accent=red aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/skchronicles/RNA-seek/ title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> CCBR/RNA-seek </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> About </a> </li> <li class=md-tabs__item> <a href=../Theory/ class=md-tabs__link> Theory </a> </li> <li class=md-tabs__item> <a href=./ class="md-tabs__link md-tabs__link--active"> Getting started </a> </li> <li class=md-tabs__item> <a href=../run/ class=md-tabs__link> Usage </a> </li> <li class=md-tabs__item> <a href=../Resources/ class=md-tabs__link> Resources </a> </li> <li class=md-tabs__item> <a href=../../general-questions/ class=md-tabs__link> FAQ </a> </li> <li class=md-tabs__item> <a href=../../license/ class=md-tabs__link> License </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="RNA-seek Documentation" class="md-nav__button md-logo" aria-label="RNA-seek Documentation" data-md-component=logo> <img src=../../assets/icons/doc-book.svg alt=logo> </a> RNA-seek Documentation </label> <div class=md-nav__source> <a href=https://github.com/skchronicles/RNA-seek/ title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> CCBR/RNA-seek </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> About </a> </li> <li class=md-nav__item> <a href=../Theory/ class=md-nav__link> Theory </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Getting started <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> Getting started </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#1-introduction class=md-nav__link> 1. Introduction </a> </li> <li class=md-nav__item> <a href=#2-setup-rna-seek class=md-nav__link> 2. Setup RNA-seek </a> <nav class=md-nav aria-label="2. Setup RNA-seek"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#21-login-to-cluster class=md-nav__link> 2.1 Login to cluster </a> </li> <li class=md-nav__item> <a href=#22-grab-an-interactive-node class=md-nav__link> 2.2 Grab an interactive node </a> </li> <li class=md-nav__item> <a href=#23-load-dependecies class=md-nav__link> 2.3 Load dependecies </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#3-building-reference-files class=md-nav__link> 3. Building Reference files </a> <nav class=md-nav aria-label="3. Building Reference files"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#31-download-references-from-gencode class=md-nav__link> 3.1 Download References from GENCODE </a> </li> <li class=md-nav__item> <a href=#32-run-build-pipeline class=md-nav__link> 3.2 Run Build pipeline </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#4-running-rna-seek class=md-nav__link> 4. Running RNA-seek </a> <nav class=md-nav aria-label="4. Running RNA-seek"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#41-dry-run-pipeline class=md-nav__link> 4.1 Dry-run pipeline </a> </li> <li class=md-nav__item> <a href=#42-run-pipeline class=md-nav__link> 4.2 Run pipeline </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_4 type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4> Usage <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Usage data-md-level=1> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Usage </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../run/ class=md-nav__link> run </a> </li> <li class=md-nav__item> <a href=../build/ class=md-nav__link> build </a> </li> <li class=md-nav__item> <a href=../cache/ class=md-nav__link> cache </a> </li> <li class=md-nav__item> <a href=../unlock/ class=md-nav__link> unlock </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../Resources/ class=md-nav__link> Resources </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_6 type=checkbox id=__nav_6> <label class=md-nav__link for=__nav_6> FAQ <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=FAQ data-md-level=1> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../general-questions/ class=md-nav__link> General Questions </a> </li> <li class=md-nav__item> <a href=../../troubleshooting/ class=md-nav__link> Troubleshooting </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../license/ class=md-nav__link> License </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/skchronicles/RNA-seek/edit/master/docs/RNA-seq/TLDR-RNA-seq.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg> </a> <h1>Getting started</h1> <h2 id=1-introduction>1. Introduction<a class=headerlink href=#1-introduction title="Permanent link">&para;</a></h2> <p>When processing RNA-sequencing data, there are often many steps that we must repeat. These are usually steps like removing adapter sequences, aligning reads against a reference genome, checking the quality of the data, and quantifying counts. RNA-seek is composed of several sub commands or convience functions to automate these repetitive steps.</p> <p>With RNA-seek, you can run your samples through our highly-reproducible pipeline, build resources for new reference genomes, and more!</p> <p>Here is a list of available rna-seek <code>sub commands</code>:</p> <ul> <li><a href=../run><strong><code>run</code></strong></a>: run the rna-seq pipeline </li> <li><a href=../build><strong><code>build</code></strong></a>: build reference files </li> <li><a href=../cache><strong><code>cache</code></strong></a>: cache remote resources locally </li> <li><a href=../unlock><strong><code>unlock</code></strong></a>: unlock a working directory </li> </ul> <blockquote> <p>This page contains information for building reference files and running the RNA-seek pipeline. For more information about each of the available sub commands, please see the <a href>usage section</a>. </p> </blockquote> <h2 id=2-setup-rna-seek>2. Setup RNA-seek<a class=headerlink href=#2-setup-rna-seek title="Permanent link">&para;</a></h2> <p><em>Estimated Reading Time: 3 Mintutes</em></p> <p>RNA-seek has two dependencies: <code>singularity</code> and <code>snakemake</code>. These dependencies can be installed by a sysadmin; however, snakemake is readily available through conda. Before running the pipeline or any of the commands below, please ensure singularity and snakemake are in your <code>$PATH</code>. Please see follow the instructions below for getting started with the RNA-seek pipeline.</p> <h3 id=21-login-to-cluster>2.1 Login to cluster<a class=headerlink href=#21-login-to-cluster title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Setup Step 0.) ssh into cluster&#39;s head node</span>
<span class=c1># example below for Biowulf cluster</span>
ssh -Y <span class=nv>$USER</span>@biowulf.nih.gov
</code></pre></div> <h3 id=22-grab-an-interactive-node>2.2 Grab an interactive node<a class=headerlink href=#22-grab-an-interactive-node title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Setup Step 1.) Please do not run RNA-seek on the head node!</span>
<span class=c1># Grab an interactive node first</span>
srun -N <span class=m>1</span> -n <span class=m>1</span> --time<span class=o>=</span><span class=m>12</span>:00:00 -p interactive --mem<span class=o>=</span>8gb  --cpus-per-task<span class=o>=</span><span class=m>4</span> --pty bash
</code></pre></div> <h3 id=23-load-dependecies>2.3 Load dependecies<a class=headerlink href=#23-load-dependecies title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Setup Step 2.) Add singularity and snakemake executables to $PATH</span>
module purge
module load singularity snakemake

<span class=c1># Setup Step 3.) Download RNA-seek and add to $PATH</span>
<span class=c1># Clone the RNA-seek repository from Github</span>
git clone https://github.com/skchronicles/RNA-seek.git
<span class=nb>export</span> <span class=nv>PATH</span><span class=o>=</span><span class=si>${</span><span class=nv>PWD</span><span class=si>}</span>/RNA-seek:<span class=si>${</span><span class=nv>PATH</span><span class=si>}</span>
</code></pre></div> <h2 id=3-building-reference-files>3. Building Reference files<a class=headerlink href=#3-building-reference-files title="Permanent link">&para;</a></h2> <p>In this example, we will start off by building reference files downloaded from <a href=https://www.gencodegenes.org/ >GENCODE</a>. We recommend downloading the <code>PRI</code> Genome FASTA file and annotation from <a href=https://www.gencodegenes.org/ >GENCODE</a>. These <code>PRI</code> reference files contain the primary chromosomes and scaffolds. We <strong>do not</strong> recommend downloading the <code>CHR</code> reference files! </p> <p>Here is more information about GENCODE's <a href=https://www.gencodegenes.org/human/release_36.html>v36 release</a> for the human reference genome.</p> <h3 id=31-download-references-from-gencode>3.1 Download References from GENCODE<a class=headerlink href=#31-download-references-from-gencode title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Build Step 0.) Please do not run RNA-seek on the head node!</span>
<span class=c1># Grab an interactive node first</span>
<span class=c1># Assumes that you have already ssh-ed into cluster</span>
srun -N <span class=m>1</span> -n <span class=m>1</span> --time<span class=o>=</span><span class=m>12</span>:00:00 -p interactive --mem<span class=o>=</span>8gb  --cpus-per-task<span class=o>=</span><span class=m>4</span> --pty bash

<span class=c1># Build Step 1.) Download the PRI Genome FASTA file for GRCh38.p13</span>
wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_36/GRCh38.primary_assembly.genome.fa.gz
gzip -d GRCh38.primary_assembly.genome.fa.gz

<span class=c1># Build Step 2.) Download the PRI release 36 annotation</span>
wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_36/gencode.v36.primary_assembly.annotation.gtf.gz
gzip -d gencode.v36.primary_assembly.annotation.gtf.gz
</code></pre></div> <h3 id=32-run-build-pipeline>3.2 Run Build pipeline<a class=headerlink href=#32-run-build-pipeline title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Build Step 3.) Load dependencies</span>
module purge
module load singularity snakemake

<span class=c1># Build Step 4.) Dry-run the build pipeline</span>
rna-seek build  --ref-fa GRCh38.primary_assembly.genome.fa <span class=se>\</span>
                --ref-name hg38 <span class=se>\</span>
                --ref-gtf gencode.v36.primary_assembly.annotation.gtf <span class=se>\</span>
                --gtf-ver <span class=m>36</span> --output /data/<span class=nv>$USER</span>/hg38_36 --dry-run


<span class=c1># Build Step 5.) Submit the build pipeline to cluster</span>
rna-seek build  --ref-fa GRCh38.primary_assembly.genome.fa <span class=se>\</span>
                --ref-name hg38 <span class=se>\</span>
                --ref-gtf gencode.v36.primary_assembly.annotation.gtf <span class=se>\</span>
                --gtf-ver <span class=m>36</span> --output /data/<span class=nv>$USER</span>/hg38_36 
</code></pre></div> <p>An email notification will be sent out when the pipeline starts and ends. Once the build pipeline completes, you can run RNA-seek with the provided test dataset. Please see the intructions below for more information.</p> <h2 id=4-running-rna-seek>4. Running RNA-seek<a class=headerlink href=#4-running-rna-seek title="Permanent link">&para;</a></h2> <p>Run RNA-seek with the reference files we built above using hg38 (GRCh38.p13) Genome FASTA file and GENCODE release 36 annotation (GTF). For more information about how the reference files we generated, please see the intructions above. You can use those instructions as a guide for building any new reference genomes in the future. </p> <h3 id=41-dry-run-pipeline>4.1 Dry-run pipeline<a class=headerlink href=#41-dry-run-pipeline title="Permanent link">&para;</a></h3> <p>Dry-run the pipeline prior to submiting the pipeline's master job. Please note that if you wish to run RNA-seek with a new dataset, you will only need to update the values provided to the <code>--input</code> and <code>--output</code> arguments (and maybe <code>--genome</code>). The <code>--input</code> argument supports globbing. If this is the first time running RNA-seek with for given dataset, the <code>--output</code> directory should <em><strong>not</strong></em> exist on your local filesystem. It will be created automatically during runtime.</p> <div class=highlight><pre><span></span><code><span class=c1># Run Step 0.) Please do not run RNA-seek on the head node!</span>
<span class=c1># Grab an interactive node first</span>
<span class=c1># Assumes that you have already ssh-ed into cluster</span>
srun -N <span class=m>1</span> -n <span class=m>1</span> --time<span class=o>=</span><span class=m>12</span>:00:00 -p interactive --mem<span class=o>=</span>8gb  --cpus-per-task<span class=o>=</span><span class=m>4</span> --pty bash

<span class=c1># Run Step 1.) Load dependencies</span>
module purge
module load singularity snakemake

<span class=c1># Run Step 2.) Dry-run the pipeline with test dataset</span>
<span class=c1># And reference genome generated in the steps above</span>
<span class=c1># Test data consists of sub sampled FastQ files </span>
rna-seek run <span class=se>\</span>
    --input RNA-seek/.tests/*.R?.fastq.gz <span class=se>\</span>
    --output /data/<span class=si>${</span><span class=nv>USER</span><span class=si>}</span>/runner_hg38_36/  <span class=se>\</span>
    --genome /data/<span class=si>${</span><span class=nv>USER</span><span class=si>}</span>/hg38_36/hg38_36.json  <span class=se>\</span>
    --mode slurm <span class=se>\</span>
    --star-2-pass-basic <span class=se>\</span>
    --dry-run
</code></pre></div> <h3 id=42-run-pipeline>4.2 Run pipeline<a class=headerlink href=#42-run-pipeline title="Permanent link">&para;</a></h3> <p>Kick off the pipeline by submiting the master job to the cluster. It is essentially the same command above without the <code>--dry-run</code> flag. </p> <div class=highlight><pre><span></span><code><span class=c1># Run Step 3.) Submit the master job</span>
<span class=c1># Runs the RNA-seek pipeline  with the </span>
<span class=c1># reference genome generated in the steps above</span>
<span class=c1># and with the test dataset</span>
rna-seek run <span class=se>\</span>
    --input RNA-seek/.tests/*.R?.fastq.gz <span class=se>\</span>
    --output /data/<span class=si>${</span><span class=nv>USER</span><span class=si>}</span>/runner_hg38_36/  <span class=se>\</span>
    --genome /data/<span class=si>${</span><span class=nv>USER</span><span class=si>}</span>/hg38_36/hg38_36.json  <span class=se>\</span>
    --mode slurm <span class=se>\</span>
    --star-2-pass-basic <span class=se>\</span>
    --dry-run
</code></pre></div> <p>An email notification will be sent out when the pipeline starts and ends.</p> <hr> <div class=md-source-date> <small> Last update: 2025-10-24 </small> </div> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top data-md-state=hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../Theory/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Theory" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Theory </div> </div> </a> <a href=../run/ class="md-footer__link md-footer__link--next" aria-label="Next: run" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> run </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-footer-copyright> <div class=md-footer-copyright__highlight> Copyright &copy; 2021 CCBR </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-footer-social> <a href=http://bioinformatics.cancer.gov target=_blank rel=noopener title=bioinformatics.cancer.gov class=md-footer-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><path d="M96 224c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm448 0c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm32 32h-64c-17.6 0-33.5 7.1-45.1 18.6 40.3 22.1 68.9 62 75.1 109.4h66c17.7 0 32-14.3 32-32v-32c0-35.3-28.7-64-64-64zm-256 0c61.9 0 112-50.1 112-112S381.9 32 320 32 208 82.1 208 144s50.1 112 112 112zm76.8 32h-8.3c-20.8 10-43.9 16-68.5 16s-47.6-6-68.5-16h-8.3C179.6 288 128 339.6 128 403.2V432c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48v-28.8c0-63.6-51.6-115.2-115.2-115.2zm-223.7-13.4C161.5 263.1 145.6 256 128 256H64c-35.3 0-64 28.7-64 64v32c0 17.7 14.3 32 32 32h65.9c6.3-47.4 34.9-87.3 75.2-109.4z"/></svg> </a> <a href=https://github.com/CCBR target=_blank rel=noopener title=github.com class=md-footer-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://hub.docker.com/orgs/nciccbr/repositories target=_blank rel=noopener title=hub.docker.com class=md-footer-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><path d="M349.9 236.3h-66.1v-59.4h66.1v59.4zm0-204.3h-66.1v60.7h66.1V32zm78.2 144.8H362v59.4h66.1v-59.4zm-156.3-72.1h-66.1v60.1h66.1v-60.1zm78.1 0h-66.1v60.1h66.1v-60.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1l-13.3-8.9zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm-78.1-72.1h-66.1v60.1h66.1v-60.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../..", "features": ["navigation.tabs", "navigation.top", "toc.integrate"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../assets/javascripts/workers/search.709b4209.min.js", "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.29db7785.min.js></script> </body> </html>