<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="RNA-seek is an open-source, reproducible, and scalable best practices pipeline for analyzing RNA-sequencing data. "><meta name=author content="Skyler Kuhn"><link rel=icon href=../../assets/favicon/favicon.ico><meta name=generator content="mkdocs-1.1.2, mkdocs-material-7.2.3"><title>run - RNA-seek Documentation</title><link rel=stylesheet href=../../assets/stylesheets/main.f7f47774.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.3f5d1f46.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback"><style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <script>function __prefix(e){return new URL("../..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script> <script>var palette=__get("__palette");if(null!==palette&&"object"==typeof palette.color)for(var key in palette.color)document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#rna-seek-run class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="RNA-seek Documentation" class="md-header__button md-logo" aria-label="RNA-seek Documentation" data-md-component=logo> <img src=../../assets/icons/doc-book.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> RNA-seek Documentation </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> run </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M7 10a2 2 0 0 1 2 2 2 2 0 0 1-2 2 2 2 0 0 1-2-2 2 2 0 0 1 2-2m10-3a5 5 0 0 1 5 5 5 5 0 0 1-5 5H7a5 5 0 0 1-5-5 5 5 0 0 1 5-5h10M7 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3h10a3 3 0 0 0 3-3 3 3 0 0 0-3-3H7z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=red data-md-color-accent=red aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/skchronicles/RNA-seek/ title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> CCBR/RNA-seek </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> About </a> </li> <li class=md-tabs__item> <a href=../Theory/ class=md-tabs__link> Theory </a> </li> <li class=md-tabs__item> <a href=../TLDR-RNA-seq/ class=md-tabs__link> Getting started </a> </li> <li class=md-tabs__item> <a href=./ class="md-tabs__link md-tabs__link--active"> Usage </a> </li> <li class=md-tabs__item> <a href=../Resources/ class=md-tabs__link> Resources </a> </li> <li class=md-tabs__item> <a href=../../general-questions/ class=md-tabs__link> FAQ </a> </li> <li class=md-tabs__item> <a href=../../license/ class=md-tabs__link> License </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="RNA-seek Documentation" class="md-nav__button md-logo" aria-label="RNA-seek Documentation" data-md-component=logo> <img src=../../assets/icons/doc-book.svg alt=logo> </a> RNA-seek Documentation </label> <div class=md-nav__source> <a href=https://github.com/skchronicles/RNA-seek/ title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> CCBR/RNA-seek </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> About </a> </li> <li class=md-nav__item> <a href=../Theory/ class=md-nav__link> Theory </a> </li> <li class=md-nav__item> <a href=../TLDR-RNA-seq/ class=md-nav__link> Getting started </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_4 type=checkbox id=__nav_4 checked> <label class=md-nav__link for=__nav_4> Usage <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Usage data-md-level=1> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Usage </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> run <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> run </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#1-about class=md-nav__link> 1. About </a> </li> <li class=md-nav__item> <a href=#2-synopsis class=md-nav__link> 2. Synopsis </a> <nav class=md-nav aria-label="2. Synopsis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#21-required-arguments class=md-nav__link> 2.1 Required Arguments </a> </li> <li class=md-nav__item> <a href=#22-analysis-options class=md-nav__link> 2.2 Analysis Options </a> </li> <li class=md-nav__item> <a href=#23-downstream-analysis-options class=md-nav__link> 2.3 Downstream analysis options </a> </li> <li class=md-nav__item> <a href=#24-orchestration-options class=md-nav__link> 2.4 Orchestration Options </a> </li> <li class=md-nav__item> <a href=#25-misc-options class=md-nav__link> 2.5 Misc Options </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#3-example class=md-nav__link> 3. Example </a> <nav class=md-nav aria-label="3. Example"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#31-biowulf class=md-nav__link> 3.1 Biowulf </a> </li> <li class=md-nav__item> <a href=#32-generic-slurm-cluster class=md-nav__link> 3.2 Generic SLURM Cluster </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../build/ class=md-nav__link> build </a> </li> <li class=md-nav__item> <a href=../cache/ class=md-nav__link> cache </a> </li> <li class=md-nav__item> <a href=../unlock/ class=md-nav__link> unlock </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../Resources/ class=md-nav__link> Resources </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_6 type=checkbox id=__nav_6> <label class=md-nav__link for=__nav_6> FAQ <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=FAQ data-md-level=1> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../general-questions/ class=md-nav__link> General Questions </a> </li> <li class=md-nav__item> <a href=../../troubleshooting/ class=md-nav__link> Troubleshooting </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../license/ class=md-nav__link> License </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/skchronicles/RNA-seek/edit/master/docs/RNA-seq/run.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg> </a> <h1 id=rna-seek-run><code>rna-seek <b>run</b></code><a class=headerlink href=#rna-seek-run title="Permanent link">&para;</a></h1> <h2 id=1-about>1. About<a class=headerlink href=#1-about title="Permanent link">&para;</a></h2> <p>The <code>rna-seek</code> executable is composed of several inter-related sub commands. Please see <code>rna-seek -h</code> for all available options.</p> <p>This part of the documentation describes options and concepts for <code>rna-seek <b>run</b></code> sub command in more detail. With minimal configuration, the <strong><code>run</code></strong> sub command enables you to start running the data processing and quality-control pipeline. </p> <p>Setting up the RNA-seek pipeline is fast and easy! In its most basic form, <code>rna-seek <b>run</b></code> only has <em>three required inputs</em>.</p> <h2 id=2-synopsis>2. Synopsis<a class=headerlink href=#2-synopsis title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code>$ rna-seek run [--help] \
            [--batch-id BATCH_ID] \
            [--groups GROUPS] [--contrasts CONTRASTS] \
            [--covariates COVARIATE [COVARIATE ...]] \
            [--call-gene-fusions] [--prokaryote] \
            [--small-rna] [--star-2-pass-basic] \
            [--dry-run] [--mode {slurm, local}] \
            [--shared-resources SHARED_RESOURCES] \
            [--singularity-cache SINGULARITY_CACHE] \
            [--sif-cache SIF_CACHE] \
            [--tmp-dir TMP_DIR] \
            [--threads THREADS] \
            --input INPUT [INPUT ...] \
            --output OUTPUT \
            --genome {hg38_48, mm10_M25, mm39_M37, custom.json}
</code></pre></div> <p>The synopsis for each command shows its parameters and their usage. Optional parameters are shown in square brackets.</p> <p>A user <strong>must</strong> provide a list of FastQ files (globbing is supported) to analyze via <code>--input</code> argument, an output directory to store results via <code>--output</code> argument and select reference genome for alignment and annotation via the <code>--genome</code> argument. If you are running the pipeline outside of Biowulf, you will need to additionally provide the the following options: <code>--shared-resources</code>, <code>--tmp-dir</code>. More information about each of these options can be found below.</p> <p>To run differential expression analyses, a groups and contrasts file must be provided. Please see the downstream analysis section below for more information about how to create these files.</p> <p>Use you can always use the <code>-h</code> option for information on a specific sub command. </p> <h3 id=21-required-arguments>2.1 Required Arguments<a class=headerlink href=#21-required-arguments title="Permanent link">&para;</a></h3> <p>Each of the following arguments are required. Failure to provide a required argument will result in a non-zero exit-code.</p> <p><code>--input INPUT [INPUT ...]</code> </p> <blockquote> <p><strong>Input FastQ file(s) to process.</strong><br> <em>type: file</em> </p> <p>One or more FastQ files can be provided. From the command-line, each FastQ file should seperated by a space. Globbing is supported! This makes selecting FastQ files easier. Input FastQ files should be gzipp-ed. The pipeline supports single-end and pair-end RNA-seq data; however, the pipeline will not process a mixture of single-end and paired-end samples together. If you have a mixture of single-end and pair-end samples to process, please process them as two seperate instances of the RNA-seek pipeline (with two seperate output directories). </p> <p><strong><em>Example:</em></strong> <code>--input .tests/*.R?.fastq.gz</code></p> </blockquote> <hr> <p><code>--output OUTPUT</code></p> <blockquote> <p><strong>Path to an output directory.</strong> <br> <em>type: path</em></p> <p>This location is where the pipeline will create all of its output files, also known as the pipeline's working directory. If the provided output directory does not exist, it will be initialized automatically.</p> <p><strong><em>Example:</em></strong> <code>--output /data/$USER/RNA_hg38</code></p> </blockquote> <hr> <p><code>--genome {hg38_48, mm10_M25, mm39_M37, custom.json}</code></p> <blockquote> <p><strong>Reference genome.</strong> <br> <em>type: string or file</em></p> <p>This option defines the reference genome for your set of samples. On Biowulf, RNA-seek does comes bundled with pre built reference files for human and mouse samples; however, it is worth noting that the pipeline does accept a custom reference genome built with the build sub command. Building a new reference genome is easy! You can create a custom reference genome with a single command. This is extremely useful when working with non-model organisms. New users can reference the documentation's <a href=../TLDR-RNA-seq/#3-building-reference-files>getting started</a> section to see how a reference genome is built. </p> <p><strong><em>Pre built Option</em></strong><br> Here is a list of available pre built genomes on Biowulf: hg38_48, mm10_M25, mm39_M37. Please see the <a href=../RNA-seq//Resources/#1-reference-genomes>resources page</a> for more information about each pre built option. It is worth noting that hg38_30 and mm10_M21 can still be selected. They have just been hidden from the command-line interface to encourage users to select newer reference genomes.</p> <p><strong><em>Custom Option</em></strong> <br> A user can also supply a custom reference genome built with the build sub command. Please supply the custom reference JSON file that was generated by the build sub command. The name of this custom reference JSON file is dependent on the values provided to the following <em>rna-seek build</em> args, <code>--ref-name REF_NAME</code> and <code>--gtf-ver GTF_VER</code>, where the name of the provided custom reference JSON file would be: <code>{REF_NAME}_{GTF_VER}.json</code>.</p> <p><strong><em>Example:</em></strong> <code>--genome hg38_48</code> <em>OR</em> <code>--genome /data/${USER}/hg38_36/hg38_36.json</code> </p> </blockquote> <h3 id=22-analysis-options>2.2 Analysis Options<a class=headerlink href=#22-analysis-options title="Permanent link">&para;</a></h3> <p><code>--call-gene-fusions</code> </p> <blockquote> <p><strong>Call gene fusions.</strong><br> <em>type: boolean</em></p> <p>A gene fusion is a genetic alteration where two different genes become joined together, forming a new hybrid gene. If this option is provided, an extra set of steps will run to detect gene fusions with arriba. These steps will conditionally run depending on whether there are reference files for running arriba. Currently, these reference files exist for <code>hg38</code>, <code>hg19</code>, and <code>mm10</code>. Please note if these reference are missing, then gene fusions will not be called. </p> <p><strong><em>Example:</em></strong> <code>--call-gene-fusions</code></p> </blockquote> <hr> <p><code>--prokaryote</code> </p> <blockquote> <p><strong>Run with prokaryotic genome alignment options.</strong><br> <em>type: boolean</em></p> <p>Prokaryotic genomes, like bacteria, do not contain introns. If provided, this option will use an optimized set of options for aligning against prokaryotic genomes. This option will force STAR to avoid spliced alignments, and it will also run STAR in a 2-pass basic mode. By default, the pipeline is setup for handling alignment against eukarytoic genomes, so this option should be provided if you are working with a prokaryotic genome. This option should not be combined with the small RNA option. </p> <p><strong><em>Example:</em></strong> <code>--prokaryote</code></p> </blockquote> <hr> <p><code>--small-rna</code> </p> <blockquote> <p><strong>Run STAR using ENCODE's recomendations for small RNA.</strong><br> <em>type: boolean</em></p> <p>This option should only be used with small RNA libraries. These are rRNA-depleted libraries that have been size selected to contain fragments shorter than 200bp. Size selection enriches for small RNA species such as miRNAs, siRNAs, or piRNAs. Also, this option should not be combined with the star 2-pass basic option. If the two options are combined, STAR will run in pass basic mode. This means that STAR will not run with ENCODE's recommendations for small RNA alignment. As so, please take caution not to combine both options together. </p> <p>Please note: This option is only supported with single-end data. </p> <p><strong><em>Example:</em></strong> <code>--small-rna</code></p> </blockquote> <hr> <p><code>--star-2-pass-basic</code> </p> <blockquote> <p><strong>Run STAR in per sample 2-pass mapping mode.</strong><br> <em>type: boolean</em></p> <p>It is recommended to use this option when processing a set of unrelated samples or when processing samples in a clinical setting. It is not adivsed to use this option for a study with multiple related samples. </p> <p>By default, the pipeline ultilizes a multi sample 2-pass mapping approach where the set of splice junctions detected across all samples are provided to the second pass of STAR. This option overrides the default behavior so each sample will be processed in a per sample two-pass basic mode. This option should not be combined with the small RNA option. If the two options are combined, STAR will run in pass basic mode.</p> <p><strong><em>Example:</em></strong> <code>--star-2-pass-basic</code></p> </blockquote> <h3 id=23-downstream-analysis-options>2.3 Downstream analysis options<a class=headerlink href=#23-downstream-analysis-options title="Permanent link">&para;</a></h3> <p>Each of the following options can optionally be provided to run differential expression and pathway (coming soon) analyses. Please note that if these options are not provided, the pipeline will still run all pre-processing and QC steps.</p> <p><code>--batch-id BATCH_ID</code></p> <blockquote> <p><strong>Batch ID for downstream analysis.</strong><br> <em>type: string</em><br> <em>default: none</em> </p> <p>This option can be provided to ensure that downstream analyses output files are not over-written between different runs of the pipeline. This can occur after updating the group file with additional covariates or dropping samples. By default, project-level files in differential expression and pathway analysis folder could get over-written between pipeline runs if this option is not provided. The output directory name for a given contrast will resolve to <code>{group1}_vs_{group2}</code> within the <strong>differential_gene_expression</strong> folder. As so, if the groups file is updated to remove samples or add additional covariates without updating the group names, it could over write the previous analyses output files. Any identifer provided to this option will be used to create a sub directory in the differential expression folder. This ensures project-level files (which are unique) will not get over written. With that being said, it is always a good idea to provide this option. A unique batch id should be provided between runs. This batch id should be composed of alphanumeric characters and it should not contain a white space or tab characters. Here is a list of valid or acceptable characters: <code>aA-Zz</code>, <code>0-9</code>, <code>-</code>, <code>_</code>.</p> <p><strong><em>Example:</em></strong> <code>--batch-id 2025_04_08</code></p> </blockquote> <hr> <p><code>--groups GROUPS</code></p> <blockquote> <p><strong>Groups file containing sample metadata.</strong> <br> <em>type: TSV file</em><br> <em>default: none</em> </p> <p>Path to a sample sheet in TSV format with important information for differential expression. This file is used to map the sample names to groups and other metadata such as covariates (optional). At a minimum, this file contains two columns containing the basename of the sample along with its group label. Additional columns can be optionally added to define covariates. To run differential expression analyses, a groups and contrasts file must be provided together. Please see the example below for more information about how to create this file.</p> <p>Here is a more detailed description of the sample sheet:<br> • <strong>Sample</strong>: This is the basename and prefix of a sample's R1 FastQ file, e.g <code>/path/WT_S4.R1.fastq.gz</code> becomes <code>WT_S4</code> in this file. Please note a sample should only be listed once in the groups file, and that there should only be a 1:1 relationship between samples and groups. This is a required column.<br> • <strong>Group</strong>: This is the group label for a given sample. A group can represent anything from an experimental condition, a treatment, timepoint, or disease state, etc. Group names must start with a letter and should then only be composed of alphanumeric characters (<code>Aa-Za,0-9</code>), periods (<code>.</code>), or underscores (<code>_</code>). Hyphens (<code>-</code>) are not allowed in group names! This is a required column.<br> • <strong>Additional Columns</strong> (optional): Any additional columns to define covariates. The column names here must be unique and follow the same naming rules as group names. These additional column names can then be provided to the --covariates option. If an additional column is provided and its is not supplied to the --covariates option, then it will only be used for exploratory plots. Please see that option for more information. </p> <p><em>Contents of example groups file:</em> <div class=highlight><pre><span></span><code>Sample    Group   Batch
WT_S1 LPS_WT  B1
WT_S2 LPS_WT  B2
WT_S3 LPS_WT  B1
WT_S4 LPS_WT  B2
WT_S5 Veh_WT  B1
WT_S6 Veh_WT  B2
WT_S7 Veh_WT  B1
WT_S8 Veh_WT  B2
KO_S9 LPS_KO  B1
KO_S10    LPS_KO  B2
KO_S11    LPS_KO  B1
KO_S12    LPS_KO  B2
KO_S13    Veh_KO  B1
KO_S14    Veh_KO  B2
KO_S15    Veh_KO  B1
KO_S16    Veh_KO  B2
</code></pre></div></p> <p><strong>where:</strong> <br> • <code>Sample</code> is the base name of each sample's input R1 fastq file without <em>.R1.fastq.gz</em> file extension.<br> • <code>Group</code> represents each sample's group name(s).<br> • Any additional columns are optional covariates. Only the <code>Sample</code> and <code>Group</code> columns are required. </p> <p><strong><em>Example:</em></strong> <code>--groups .tests/groups.tsv</code></p> </blockquote> <hr> <p><code>--contrasts CONTRASTS</code></p> <blockquote> <p><strong>Contrasts file containing comparisons to make.</strong> <br> <em>type: TSV file</em><br> <em>default: none</em> </p> <p>This tab-delimited file is used to setup comparisons within different groups of samples. Please see the --groups option above for more information about how to define groups for a set of samples. In its most basic form, this file consists of two columns containing the names of two group to compare. The names defined in this file must also exist in the groups file. The second column represents the baseline group. Complex comparisons can be made. The last line in the example contrasts file below contains a difference of differences (diff-diff) comparison. For diff-diff comparisons, please wrap each diff in parentheses. To perform differential expression analysis, a groups and contrast file must be provided. Please note that unlike the groups file, the contrasts file does not contain a header line.</p> <p><em>Contents of example contrasts file:</em> <div class=highlight><pre><span></span><code>LPS_KO    Veh_KO
LPS_WT    Veh_WT
LPS_KO    LPS_WT
Veh_KO    Veh_WT
(LPS_KO-Veh_KO)   (LPS_WT-Veh_WT)
</code></pre></div></p> <p><strong>where:</strong><br> • Any groups listed in this file must exist in the <code>--groups</code> file.<br> • 2<sup>nd</sup> column represents the baseline of the comparison, i.e <em>Veh_KO</em> on the first line in the example above. </p> <p><strong><em>Example:</em></strong> <code>--contrasts .tests/contrasts.tsv</code></p> </blockquote> <hr> <p><code>--covariates COVARIATE [COVARIATE ...]</code></p> <blockquote> <p><strong>Covariates for differential analyses.</strong><br> <em>type: string</em><br> <em>default: none</em> </p> <p>This option allows the user to define and adjust for any covariates that may be present. Covariates are known biological (i.e sex, disease stage, high-/low-BMI) or technical sources (i.e. batch, high-/low-RIN) of variation that can confound the results of differential expression analysis. If you have more than one covariate, please provide a <em>space seperated list</em> of covariates (i.e <code>--covariates Batch Site</code>) to this option. Covariates must match the header names of any <strong>Additional Columns</strong> in the groups file. Covariate values should be encoded as categorical variables. Numerical covariates are currently not supported.</p> <p><strong><em>Example:</em></strong> <code>--covariates Batch</code></p> </blockquote> <h3 id=24-orchestration-options>2.4 Orchestration Options<a class=headerlink href=#24-orchestration-options title="Permanent link">&para;</a></h3> <p>Each of the following arguments are optional and do not need to be provided. </p> <p><code>--dry-run</code> </p> <blockquote> <p><strong>Dry run the pipeline.</strong><br> <em>type: boolean</em></p> <p>Displays what steps in the pipeline remain or will be run. Does not execute anything!</p> <p><strong><em>Example:</em></strong> <code>--dry-run</code></p> </blockquote> <hr> <p><code>--mode {slurm,local}</code> </p> <blockquote> <p><strong>Execution Method.</strong> <em>type: string</em><br> <em>default: slurm</em> </p> <p>Execution Method. Defines the mode or method of execution. Vaild mode options include: slurm or local. </p> <p><strong><em>local</em></strong><br> Local executions will run serially on compute instance. This is useful for testing, debugging, or when a users does not have access to a high performance computing environment. If this option is not provided, it will default to a local execution mode. </p> <p><strong><em>slurm</em></strong> <br> The slurm execution method will submit jobs to a cluster using a slurm + singularity backend. This method will automatically submit the master job to the cluster. It is recommended running RNA-seek in this mode as execution will be significantly faster in a distributed environment.</p> <p><strong><em>Example:</em></strong> <code>--mode slurm</code></p> </blockquote> <hr> <p><code>--shared-resources SHARED_RESOURCES</code> </p> <blockquote> <p><strong>Local path to shared resources.</strong><br> <em>type: path</em></p> <p>The pipeline uses a set of shared reference files that can be re-used across reference genomes. These currently include reference files for kraken and FQScreen. These reference files can be downloaded with the build sub command's <code>--shared-resources</code> option. With that being said, these files only need to be downloaded once. We recommend storing this files in a shared location on the filesystem that other people can access. If you are running the pipeline on Biowulf, you do NOT need to download these reference files! They already exist on the filesystem in a location that anyone can acceess; however, if you are running the pipeline on another cluster or target system, you will need to download the shared resources with the build sub command, and you will need to provide this option every time you run the pipeline. Please provide the same path that was provided to the build sub command's --shared-resources option. Again, if you are running the pipeline on Biowulf, you do NOT need to provide this option. For more information about how to download shared resources, please reference the build sub command's <code>--shared-resources</code> option.</p> <p><strong><em>Example:</em></strong> <code>--shared-resources /data/shared/rna-seek</code></p> </blockquote> <hr> <p><code>--singularity-cache SINGULARITY_CACHE</code> </p> <blockquote> <p><strong>Overrides the $SINGULARITY_CACHEDIR environment variable.</strong><br> <em>type: path</em><br> <em>default: <code>--output OUTPUT/.singularity</code></em></p> <p>Singularity will cache image layers pulled from remote registries. This ultimately speeds up the process of pull an image from DockerHub if an image layer already exists in the singularity cache directory. By default, the cache is set to the value provided to the <code>--output</code> argument. Please note that this cache cannot be shared across users. Singularity strictly enforces you own the cache directory and will return a non-zero exit code if you do not own the cache directory! See the <code>--sif-cache</code> option to create a shareable resource. </p> <p><strong><em>Example:</em></strong> <code>--singularity-cache /data/$USER/.singularity</code></p> </blockquote> <hr> <p><code>--sif-cache SIF_CACHE</code></p> <blockquote> <p><strong>Path where a local cache of SIFs are stored.</strong><br> <em>type: path</em> </p> <p>Uses a local cache of SIFs on the filesystem. This SIF cache can be shared across users if permissions are set correctly. If a SIF does not exist in the SIF cache, the image will be pulled from Dockerhub and a warning message will be displayed. The <code>rna-seek cache</code> subcommand can be used to create a local SIF cache. Please see <code>rna-seek cache</code> for more information. This command is extremely useful for avoiding DockerHub pull rate limits. It also remove any potential errors that could occur due to network issues or DockerHub being temporarily unavailable. We recommend running RNA-seek with this option when ever possible.</p> <p><strong><em>Example:</em></strong> <code>--sif-cache /data/$USER/SIFs</code></p> </blockquote> <hr> <p><code>--tmp-dir TMP_DIR</code> </p> <blockquote> <p><strong>Path on the file system for writing temporary files.</strong><br> <em>type: path</em><br> <em>default: <code>/lscratch/$SLURM_JOBID</code></em></p> <p>This is a path on the file system for writing temporary output files. By default, the temporary directory is set to '/lscratch/$SLURM_JOBID' for backwards compatibility with the NIH's Biowulf cluster; however, if you are running the pipeline on another cluster, this option will need to be specified. Ideally, this path should point to a dedicated location on the filesystem for writing tmp files. On many systems, this location is set to somewhere in /scratch. If you need to inject a variable into this string that should NOT be expanded, please quote this options value in single quotes. Again, if you are running the pipeline on Biowulf, you do NOT need to provide this option. </p> <p><strong><em>Example:</em></strong> <code>--tmp-dir /cluster_scratch/$USER/</code></p> </blockquote> <hr> <p><code>--threads THREADS</code> </p> <blockquote> <p><strong>Max number of threads for each process.</strong><br> <em>type: int</em><br> <em>default: 2</em></p> <p>Max number of threads for each process. This option is more applicable when running the pipeline with <code>--mode local</code>. It is recommended setting this vaule to the maximum number of CPUs available on the host machine.</p> <p><strong><em>Example:</em></strong> <code>--threads 12</code></p> </blockquote> <h3 id=25-misc-options>2.5 Misc Options<a class=headerlink href=#25-misc-options title="Permanent link">&para;</a></h3> <p>Each of the following arguments are optional and do not need to be provided. </p> <p><code>-h, --help</code> </p> <blockquote> <p><strong>Display Help.</strong><br> <em>type: boolean</em></p> <p>Shows command's synopsis, help message, and an example command</p> <p><strong><em>Example:</em></strong> <code>--help</code></p> </blockquote> <h2 id=3-example>3. Example<a class=headerlink href=#3-example title="Permanent link">&para;</a></h2> <h3 id=31-biowulf>3.1 Biowulf<a class=headerlink href=#31-biowulf title="Permanent link">&para;</a></h3> <p>On Biowulf getting started with the pipeline is fast and easy! The pipeline comes bundled with pre-built human and mouse reference genomes. In the example below, we will use the pre-built human reference genome. </p> <div class=highlight><pre><span></span><code><span class=c1># Step 0.) Grab an interactive node (do not run on head node)</span>
srun -N <span class=m>1</span> -n <span class=m>1</span> --time<span class=o>=</span><span class=m>12</span>:00:00 -p interactive --mem<span class=o>=</span>8gb  --cpus-per-task<span class=o>=</span><span class=m>4</span> --pty bash
module purge
module load singularity snakemake

<span class=c1># Step 1.) Dry run pipeline with provided test data</span>
./rna-seek run --input .tests/*.R?.fastq.gz <span class=se>\</span>
               --output /data/<span class=nv>$USER</span>/RNA_hg38 <span class=se>\</span>
               --genome hg38_48 <span class=se>\</span>
               --mode slurm <span class=se>\</span>
               --star-2-pass-basic <span class=se>\</span>
               --sif-cache /data/OpenOmics/SIFs/ <span class=se>\</span>
               --dry-run

<span class=c1># Step 2.) Run RNA-seek pipeline</span>
<span class=c1># The slurm mode will submit jobs to the cluster.</span>
<span class=c1># It is recommended running rna-seek in this mode.</span>
./rna-seek run --input .tests/*.R?.fastq.gz <span class=se>\</span>
               --output /data/<span class=nv>$USER</span>/RNA_hg38 <span class=se>\</span>
               --genome hg38_48 <span class=se>\</span>
               --mode slurm <span class=se>\</span>
                --sif-cache /data/OpenOmics/SIFs/ <span class=se>\</span>
               --star-2-pass-basic
</code></pre></div> <h3 id=32-generic-slurm-cluster>3.2 Generic SLURM Cluster<a class=headerlink href=#32-generic-slurm-cluster title="Permanent link">&para;</a></h3> <p>Running the pipeline outside of Biowulf is easy; however, there are a few extra steps you must first take. Before getting started, you will need to <a href=../TLDR-RNA-seq/#3-building-reference-files>build</a> reference files for the pipeline. Please note when running the build sub command for the first time, you will also need to provide the <code>--shared-resources</code> option. This option will download our kraken2 database and bowtie2 indices for FastQ Screen. The path provided to this option should be provided to the <code>--shared-resources</code> option of the run sub command. Next, you will also need to provide a path to write temporary output files via the <code>--tmp-dir</code> option. We also recommend providing a path to a SIF cache. You can cache software containers locally with the <a href=./RNA-seq/cache/ >cache</a> sub command. </p> <div class=highlight><pre><span></span><code><span class=c1># Step 0.) Grab an interactive node (do not run on head node)</span>
srun -N <span class=m>1</span> -n <span class=m>1</span> --time<span class=o>=</span><span class=m>2</span>:00:00 -p interactive --mem<span class=o>=</span>8gb  --cpus-per-task<span class=o>=</span><span class=m>4</span> --pty bash
<span class=c1># Add snakemake and singularity to $PATH,</span>
<span class=c1># This step may vary across clusters, you</span>
<span class=c1># can reach out to a sys admin if snakemake</span>
<span class=c1># and singularity are not installed.</span>
module purge
module load singularity snakemake

<span class=c1># Step 1.) Dry run pipeline with provided test data</span>
./rna-seek run --input .tests/*.R?.fastq.gz <span class=se>\</span>
               --output /data/<span class=nv>$USER</span>/RNA_hg38 <span class=se>\</span>
               --genome /data/<span class=nv>$USER</span>/hg38_36/hg38_36.json <span class=se>\</span>
               --mode slurm <span class=se>\</span>
               --sif-cache /data/<span class=nv>$USER</span>/cache <span class=se>\</span>
               --star-2-pass-basic <span class=se>\</span>
               --shared-resources /data/shared/rna-seek <span class=se>\</span>
               --tmp-dir /cluster_scratch/<span class=nv>$USER</span>/ <span class=se>\</span>
               --dry-run

<span class=c1># Step 2.) Run RNA-seek pipeline</span>
<span class=c1># The slurm mode will submit jobs to the cluster.</span>
<span class=c1># It is recommended running rna-seek in this mode.</span>
./rna-seek run --input .tests/*.R?.fastq.gz <span class=se>\</span>
               --output /data/<span class=nv>$USER</span>/RNA_hg38 <span class=se>\</span>
               --genome /data/<span class=nv>$USER</span>/hg38_36/hg38_36.json <span class=se>\</span>
               --mode slurm <span class=se>\</span>
               --sif-cache /data/<span class=nv>$USER</span>/cache <span class=se>\</span>
               --star-2-pass-basic <span class=se>\</span>
               --shared-resources /data/shared/rna-seek <span class=se>\</span>
               --tmp-dir /cluster_scratch/<span class=nv>$USER</span>/ <span class=se>\</span>
               --dry-run
</code></pre></div> <hr> <div class=md-source-date> <small> Last update: 2025-11-20 </small> </div> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top data-md-state=hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../TLDR-RNA-seq/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Getting started" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Getting started </div> </div> </a> <a href=../build/ class="md-footer__link md-footer__link--next" aria-label="Next: build" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> build </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-footer-copyright> <div class=md-footer-copyright__highlight> Copyright &copy; 2021 CCBR </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-footer-social> <a href=http://bioinformatics.cancer.gov target=_blank rel=noopener title=bioinformatics.cancer.gov class=md-footer-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><path d="M96 224c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm448 0c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm32 32h-64c-17.6 0-33.5 7.1-45.1 18.6 40.3 22.1 68.9 62 75.1 109.4h66c17.7 0 32-14.3 32-32v-32c0-35.3-28.7-64-64-64zm-256 0c61.9 0 112-50.1 112-112S381.9 32 320 32 208 82.1 208 144s50.1 112 112 112zm76.8 32h-8.3c-20.8 10-43.9 16-68.5 16s-47.6-6-68.5-16h-8.3C179.6 288 128 339.6 128 403.2V432c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48v-28.8c0-63.6-51.6-115.2-115.2-115.2zm-223.7-13.4C161.5 263.1 145.6 256 128 256H64c-35.3 0-64 28.7-64 64v32c0 17.7 14.3 32 32 32h65.9c6.3-47.4 34.9-87.3 75.2-109.4z"/></svg> </a> <a href=https://github.com/CCBR target=_blank rel=noopener title=github.com class=md-footer-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://hub.docker.com/orgs/nciccbr/repositories target=_blank rel=noopener title=hub.docker.com class=md-footer-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><path d="M349.9 236.3h-66.1v-59.4h66.1v59.4zm0-204.3h-66.1v60.7h66.1V32zm78.2 144.8H362v59.4h66.1v-59.4zm-156.3-72.1h-66.1v60.1h66.1v-60.1zm78.1 0h-66.1v60.1h66.1v-60.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1l-13.3-8.9zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm-78.1-72.1h-66.1v60.1h66.1v-60.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../..", "features": ["navigation.tabs", "navigation.top", "toc.integrate"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../assets/javascripts/workers/search.709b4209.min.js", "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.29db7785.min.js></script> </body> </html>